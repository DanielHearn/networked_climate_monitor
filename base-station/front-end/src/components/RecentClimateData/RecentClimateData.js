import { formatClimateData } from './../../static/helpers'

// Displays a list of climate data
export default {
  name: 'recent-climate-data',
  props: {
    recentClimateData: {
      type: Array,
      default: [],
      required: true
    },
    trends: {
      type: Object,
      default: () => {
        return {}
      },
      required: false
    },
    previousClimateData: {
      type: Array,
      default: () => {
        return []
      },
      required: false
    },
    temperatureUnit: {
      type: String,
      default: 'c',
      required: true
    }
  },
  computed: {
    // Format climate data
    climateData: function() {
      const recentClimateData = this.recentClimateData
      const climateData = []
      for (let data of recentClimateData) {
        const modifiedData = data
        const lowercaseType = data.type.toLowerCase()

        if (data.type === 'Temperature') {
          modifiedData.unit = this.temperatureUnit
        }

        if (this.trends && this.trends[lowercaseType]) {
          modifiedData.oneDayHigh = formatClimateData(
            modifiedData.type,
            this.trends[lowercaseType]['1_day']['high'],
            modifiedData.unit
          )
          modifiedData.oneDayLow = formatClimateData(
            modifiedData.type,
            this.trends[lowercaseType]['1_day']['low'],
            modifiedData.unit
          )
        }

        if (this.previousClimateData) {
          const previousData = this.previousClimateData.filter(
            climateData => climateData.type === data.type
          )[0]
          if (previousData) {
            const previousValue = previousData.value
            if (previousValue > data.value) {
              modifiedData.trendDirection = 'Down'
              modifiedData.trendDirectionIcon = 'arrow_downward'
            } else {
              modifiedData.trendDirection = 'Up'
              modifiedData.trendDirectionIcon = 'arrow_upward'
            }
            modifiedData.previousValue = formatClimateData(
              modifiedData.type,
              Math.abs(previousValue - data.value),
              modifiedData.unit
            )
          }
        }

        modifiedData.formattedText = formatClimateData(
          modifiedData.type,
          modifiedData.value,
          modifiedData.unit
        )
        climateData.push(modifiedData)
      }
      return climateData
    }
  }
}
