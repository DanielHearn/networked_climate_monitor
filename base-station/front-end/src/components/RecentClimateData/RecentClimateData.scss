@import "./../../scss/variables.scss";

.recent-data-list {
  list-style: none;
  padding: 0;
  margin-top: 0;
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  li {
    display: flex;
    flex-direction: column;
    justify-content: space-around;
    //width: 7em;
    //height: 4em;
    padding: 0.5em 0.75em;
    background: $white;
    margin-right: 1em;
    margin-bottom: 1em;
    text-align: left;
    box-shadow: #e8e8e8 2px 3px 3px;
    p {
      margin: 0;
    }
    &:first-child {
      margin-left: 0;
    }
  }
}

.mobile {
  .recent-data-list {
    display: grid;
    grid-template-columns: repeat(2, 50%);
    width: 100%;
    li {
      height: initial;
      width: initial;
    }
  }
}