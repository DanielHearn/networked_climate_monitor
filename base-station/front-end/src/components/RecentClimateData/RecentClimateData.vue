<template>
  <ul class="recent-data-list">
    <li v-for="(data, index) in climateData" :key="index">
      <p class="text">{{ data.type }}</p>
      <p class="sub-heading">
        {{ data.formattedText }}
      </p>
      <p v-if="data.previousValue && data.trendDirection" class="trend">
        <i class="material-icons">{{ data.trendDirectionIcon }}</i>
        {{ data.previousValue }}
      </p>
      <p v-if="data.oneDayHigh" class="text one-day-high">
        High:
        {{ data.oneDayHigh }}
      </p>
      <p v-if="data.oneDayLow" class="text one-day-low">
        Low: {{ data.oneDayLow }}
      </p>
    </li>
  </ul>
</template>

<script src="./RecentClimateData.js"></script>

<style lang="scss" src="./RecentClimateData.scss"></style>
