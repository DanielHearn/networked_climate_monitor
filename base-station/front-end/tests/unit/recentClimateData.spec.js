import Vuex from 'vuex'
import { shallowMount } from '@vue/test-utils'
import RecentClimateData from './../../src/components/RecentClimateData/RecentClimateData.vue'

describe('RecentClimateData.vue', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(RecentClimateData, {
      propsData: {
        recentClimateData: [
          {
            type: 'Temperature',
            value: 23.45,
            unit: 'c'
          },
          {
            type: 'Humidity',
            value: 56.34,
            unit: '%'
          }
        ],
        temperatureUnit: 'c'
      }
    })
  })

  afterEach(() => {
    wrapper.destroy()
    jest.resetAllMocks()
  })

  it('climateData', async () => {
    expect(wrapper.vm.climateData[0].formattedText).toBe('23.45Â°C')
    expect(wrapper.vm.climateData[1].formattedText).toBe('56.34%')

    wrapper.setProps({ climateData: [] })
    expect(wrapper.findAll('li').length).toBe(2)
  })

  it('has the expected html structure', () => {
    expect(wrapper.element).toMatchSnapshot()
  })
})
