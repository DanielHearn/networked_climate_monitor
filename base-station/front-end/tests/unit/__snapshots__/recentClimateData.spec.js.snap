// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecentClimateData.vue has the expected html structure 1`] = `
<ul
  class="recent-data-list"
>
  <li>
    <p
      class="text"
    >
      Temperature
    </p>
     
    <p
      class="sub-heading"
    >
      
      23.45Â°C
    
    </p>
     
    <!---->
     
    <!---->
     
    <!---->
  </li>
  <li>
    <p
      class="text"
    >
      Humidity
    </p>
     
    <p
      class="sub-heading"
    >
      
      56.34%
    
    </p>
     
    <!---->
     
    <!---->
     
    <!---->
  </li>
</ul>
`;
